backend/server/db/helpers/models.py

```diff
@@ -0,0 +1,115 @@
+from typing import Dict, List, Literal, NewType, Optional, Union
+from uuid import UUID
+from pydantic import BaseModel
+
+#
+# NewTypes
+#
+SessionID = NewType('SessionID', UUID)
+SessionToken = NewType('SessionToken', str)
+RefreshToken = NewType('RefreshToken', str)
+
+#
+# User Storage Models
+#
+class UserDegreeStorage(BaseModel):
+    programCode: str
+    specs: List[str]
+
+
+class UserCourseStorage(BaseModel):
+    code: str
+    suppressed: bool
+    mark: Union[Literal['SY', 'FL', 'PS', 'CR', 'DN', 'HD'], int, None]
+    uoc: int
+    ignoreFromProgression: bool
+
+# TODO-OLLI(pm): https://docs.pydantic.dev/latest/concepts/models/#rootmodel-and-custom-root-types
+# we can make this a RootModel or use a TypeAdapter if we want stronger type checks here
+type UserCoursesStorage = Dict[str, UserCourseStorage]
+
+class YearTerm(BaseModel):
+    Y: int
+    T: int
+
+class PlannerYear(BaseModel):
+    T0: List[str]
+    T1: List[str]
+    T2: List[str]
+    T3: List[str]
+
+class UserPlannerStorage(BaseModel):
+    unplanned: List[str]
+    startYear: int
+    isSummerEnabled: bool
+    mostRecentPastTerm: YearTerm
+    years: List[PlannerYear]
+    lockedTerms: Dict[str, bool]
+
+class _BaseUserStorage(BaseModel):
+    # NOTE: could also put uid here if we want
+    guest: bool
+
+class UserStorage(_BaseUserStorage):
+    setup: Literal[True] = True
+    degree: UserDegreeStorage
+    courses: UserCoursesStorage
+    planner: UserPlannerStorage
+
+class NotSetupUserStorage(_BaseUserStorage):
+    setup: Literal[False] = False
+
+# https://github.com/pydantic/pydantic/issues/1223#issuecomment-1152323275
+class PartialUserStorage(BaseModel):
+    # for batch user storage update
+    # TODO-OLLI(pm): remove this and use keyword args since they cannot be None
+    degree: Optional[UserDegreeStorage] = None
+    courses: Optional[UserCoursesStorage] = None
+    planner: Optional[UserPlannerStorage] = None
+
+#
+# Session Token Models (redis)
+#
+class SessionTokenInfoModel(BaseModel):
+    sid: SessionID              # id of the session behind this token
+    uid: str                    # user who owns the session
+    exp: int                    # time of expiry, will be replaced with a TTL on the cache
+
+#
+# Refresh Token Models (mongo)
+#
+class RefreshTokenInfoModel(BaseModel):
+    # object stored against the refresh token
+    sid: SessionID              # id of the session behind this token
+    expires_at: int             # time of expiry, will be replaced with a TTL on the cache
+
+#
+# Session Models (mongo)
+#
+# TODO-OLLI(pm): figure out if we want to actually annotate id token type so no casts needed
+class SessionOIDCInfoModel(BaseModel):
+    access_token: str           # most recent access token
+    raw_id_token: str           # most recent id token string
+    refresh_token: str          # most recent refresh token
+    validated_id_token: dict    # most recent valid id token object
```

Comment:
I do actually have strong types for this in the oidc packages...

My reasoning at the time for doing this was firstly that the expected fields can change afaik, so I didn't want to put strict validation on it in the mongo validations, thus neither here since these models are meant to be representation of the database contents. I also felt like due to this, it could couple the session abstraction to tightly with our expected oidc config. (like ideally the database handling just knows its a dict, then its the oidc layer is responsible for knowing its shape)

BUT that being said, doing this raw dict typing leads to casts being dropped around the place, so I am not sure I agree with my reasoning anymore. 
AND there are minimum set of fields that should always exist, so maybe I could figure out some base level of typing atleast and let there be extra fields. 
ALSO maybe since this session shape is specifically for the csesoc fedauth, that coupling is actually not a bad thing. Worst case, the fields change, we see some validation errors, change our validation shape, and drop all the old sessions which is completely fine.

not too sure tho, thoughts?

Reply:
If we couple this to the CSE fedauth, then we can get validation failures in testing

Reply:
good point, i will change